src/audio/igo_nr/igo_nr.c

```diff
@@ -750,95 +670,104 @@ static void igo_nr_lib_init(struct comp_dev *dev)
+#if CONFIG_IPC_MAJOR_4
+static void igo_nr_ipc4_params(struct processing_module *mod)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-	int32_t ret;
+	struct sof_ipc_stream_params *params = mod->stream_params;
+	struct comp_buffer *sinkb, *sourceb;
+	struct comp_dev *dev = mod->dev;
+
+	ipc4_base_module_cfg_to_stream_params(&mod->priv.cfg.base_cfg, params);
+	component_set_nearest_period_frames(dev, params->rate);
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	ipc4_update_buffer_format(sourceb, &mod->priv.cfg.base_cfg.audio_fmt);
+
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	ipc4_update_buffer_format(sinkb, &mod->priv.cfg.base_cfg.audio_fmt);
+}
+#endif /* CONFIG_IPC_MAJOR_4 */
+
+static void igo_nr_set_alignment(struct audio_stream *source, struct audio_stream *sink)
+{
+	/* Currently no optimizations those would use wider loads and stores */
+	audio_stream_init_alignment_constants(1, IGO_FRAME_SIZE, source);
+	audio_stream_init_alignment_constants(1, IGO_FRAME_SIZE, sink);
+}
+
+static int32_t igo_nr_prepare(struct processing_module *mod,
+			      struct sof_source **sources, int num_of_sources,
+			      struct sof_sink **sinks, int num_of_sinks)
+{
+	struct comp_buffer *sourceb, *sinkb;
+	struct comp_dev *dev = mod->dev;
+	struct comp_data *cd = module_get_private_data(mod);
+	//int ret;
 
 	comp_dbg(dev, "igo_nr_prepare()");
 
-	igo_nr_set_igo_params(dev);
+#if CONFIG_IPC_MAJOR_4
+	igo_nr_ipc4_params(mod);
+#endif
 
-	igo_nr_lib_init(dev);
+	// TODO: This fails, needs some fix
+	//ret = igo_nr_check_params(mod);
+	//if (ret)
+	//	return ret;
 
-	comp_dbg(dev, "post igo_nr_lib_init");
-	igo_nr_print_config(dev);
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	igo_nr_set_alignment(&sourceb->stream, &sinkb->stream);
 
-	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
-	if (ret < 0)
-		return ret;
+	igo_nr_set_igo_params(mod);
+
+	igo_nr_lib_init(mod);
 
-	if (ret == COMP_STATUS_STATE_ALREADY_SET)
-		return PPL_STATUS_PATH_STOP;
+	comp_dbg(dev, "post igo_nr_lib_init");
+	igo_nr_print_config(mod);
 
 	/* Clear in/out buffers */
 	memset(cd->in, 0, IGO_NR_IN_BUF_LENGTH * sizeof(int16_t));
 	memset(cd->out, 0, IGO_NR_OUT_BUF_LENGTH * sizeof(int16_t));
 
-	/* Default NR on */
+	/* Default NR on
+	 *
+	 * Note: There is a race condition with this switch control set and kernel set
+	 * ALSA switch control if such is defined in topology. This set overrides kernel
+	 * SOF driver set for control since it happens just after component init. The
+	 * user needs to re-apply the control to get expected operation. The owner of
+	 * this component should check if this is desired operation. A possible fix would
+	 * be set here only if kernel has not applied the switch control.
+	 */
 	cd->process_enable[cd->config.active_channel_idx] = true;
```

Comment:
@fuyuntsuo Can you please check if this needs a fix.

Reply:
Lets keep any existing assumptions, we can always come back and address if newer versions of IGO need the fix.